### PR TITLE
Make false, null and true to be recognized as constants in PHP

### LINKED ISSUE #1693: false, null and true must be recognized as constants in PHP
### `null` is not a keyword

`null` is a case­‑insensitive predefined constant (see <https://secure.php.net/manual/en/reserved.constants.php>).

```php
<?php
echo (defined('NULL') && defined('null')) ? 'Defined' : 'Not defined';
// Prints “Defined”
```

As can be seen in the table below, misplaced `null` causes a _compile_ error, whereas keywords (or language constructs) cause _parse_ errors.

| Naming    | with any keyword   | with `null`        |
| --------- | ------------------ | ------------------ |
| class     | `E_PARSE`          | `E_COMPILE_ERROR`  |
| constant  | `E_PARSE` ⁠*        | `E_NOTICE`         |
| function  | `E_PARSE`          | :heavy_check_mark: |
| method    | :heavy_check_mark: | :heavy_check_mark: |
| namespace | `E_PARSE`          | :heavy_check_mark: |
| trait     | `E_PARSE`          | `E_COMPILE_ERROR`  |
| variable  | :heavy_check_mark: | :heavy_check_mark: |

* ⁠Fails on the direct access, i. e., without `constant()`.

So I think Prism should categorize `null` as a constant, not as a keyword. This is unlikely to adversely affect backward compatibility, because `constant` is also a well­‑known class, which is already supported by most of the themes.

### `false` and `true` are also constants

`false` and `true` are also case­‑insensitive predefined constants and have all the same features that `null` has. Thus, they should get the same additional classes.

### Solution

Remove `null` from the `keyword` regex. Add the following code:

```javascript
Prism.languages.insertBefore('php', 'class-name', {
	'null': {
		pattern: /\b(?:null)\b/i,
		alias: 'constant predefined-constant'
	}
});
```

Redefine the `boolean` regex, which is inherited from `clike`:

```javascript
Prism.languages.insertBefore('php', 'class-name', {
	'boolean': {
		pattern: /\b(?:true|false)\b/i,
		alias: 'constant predefined-constant'
	}
});
```

#### `false` and `true`

| Result   | CSS classes                                  |
| -------- | -------------------------------------------- |
| Current  | `token boolean`                              |
| Expected | `token boolean constant predefined-constant` |

#### `null`

| Result   | CSS classes                                  |
| -------- | -------------------------------------------- |
| Current  | `token keyword`                              |
| Expected | `token null constant predefined-constant`    |

### PR DESCRIPTION
**Pull request by volado**
Resolves #1693.

### CODE REVIEW

File: components/prism-php.js [hunk 1]
```diff
@@ -28,6 +28,13 @@
 		}
 	});
 
+	Prism.languages.insertBefore('php', 'class-name', {
+		'boolean': {
```
[the remark this answers is not in the document]
Comment (mAAdhaTTah) [reply components/prism-php.js]: I will say, conceptually, Prism is not a strict linter. My understanding is `class true` in PHP would be a syntax error, so I'd be more concerned here w/ reducing the amount of code we ship than getting the highlighting exactly right.

[the remark this answers is not in the document]
Comment (volado) [reply components/prism-php.js]: Yes, I can keep it simple :)

```javascript
Prism.languages.php = Prism.languages.extend('clike', {
	// ...
	'constant': [
		{
			pattern: /\b[A-Z_][A-Z0-9_]*\b/
		},
		{
			pattern: /\b(?:null)\b/i
		}
	],
	'boolean': {
		pattern: /\b(?:false|true)\b/i,
		alias: 'constant'
	},
	// ...
});
```

`false` and `true` get `token boolean constant`.

`null` loses `keyword` and gets `token constant`.

Is it better?

[the remark this answers is not in the document]
Comment (volado) [reply components/prism-php.js]: > `class true` in PHP would be a syntax error

Yes.

> I will say, conceptually, Prism is not a strict linter.

Right. I definitely tried to make it lint. I’ll fix that.

[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-php.js]: Yes, that definitely better.

Just two small nits:

1. Please don't write `{ pattern: /.../ }`. Just use the regex directly.
2. Please position `boolean` before `constant`. `constant` after `boolean` creates the false impression that `boolean` is matched after `constant`.

**Comment by RunDevelopment** [comment]
Thank you for contributing!

**Comment by volado** [comment]
Great! Thank you very much for your help!

**Pull request merged**